### PR TITLE
Describe the behavior with `limit_*_batches=1|1.0`

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

The behavior for `limit_*_batches` is wildly different when the user passes `1` vs `1.0`. This can be a footgun.
This PR aims to increase transparency by showing a message clarifying the behavior.

In reaction to https://twitter.com/jxmnop/status/1494007900494303243 cc @borda @jxmorris12

Proposed by @williamFalcon 

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [n/a] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/loops/fit_loop.py [hunk 1]
```diff
@@ -263,6 +263,7 @@ def advance(self) -> None:  # type: ignore[override]
         log.detail(f"{self.__class__.__name__}: advancing loop")
         assert self.trainer.train_dataloader is not None
         dataloader = self.trainer.strategy.process_dataloader(self.trainer.train_dataloader)
+        assert self._data_fetcher is not None
```
Comment (carmocca) [inline pytorch_lightning/loops/fit_loop.py:266]: These seemingly unrelated type check changes are required by mypy...

**Comment by jxmorris12** [comment]
thank you!!!!!

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -2640,7 +2642,15 @@ def terminate_on_nan(self, val: bool) -> None:
...
 
 
-def _determine_batch_limits(batches: Union[int, float], name: str) -> Union[int, float]:
+def _determine_batch_limits(batches: Optional[Union[int, float]], name: str) -> Union[int, float]:
+    if batches is None:
+        # batches is optional to know if the user passed a value so that we can show the above info messages only to the
+        # users that set a value explicitly
+        return 1.0
+    if isinstance(batches, int) and batches == 1:
+        log.info(f"`Trainer({name}=1)` was configured so 1 batch per epoch will be used.")
```
[the remark this answers is not in the document]
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py]: Done!

File: pytorch_lightning/trainer/trainer.py [hunk 3]
```diff
@@ -2640,7 +2642,30 @@ def terminate_on_nan(self, val: bool) -> None:
         self._terminate_on_nan = val  # : 212
 
 
-def _determine_batch_limits(batches: Union[int, float], name: str) -> Union[int, float]:
+def _determine_batch_limits(batches: Optional[Union[int, float]], name: str) -> Union[int, float]:
+    if batches is None:
+        # batches is optional to know if the user passed a value so that we can show the above info messages only to the
+        # users that set a value explicitly
+        return 1.0
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2649]: given these are now optional on the input, does it make sense to later on refactor the loops to only check these if they're set?

File: pytorch_lightning/trainer/trainer.py [hunk 4]
```diff
@@ -2640,7 +2642,30 @@ def terminate_on_nan(self, val: bool) -> None:
...
+            message = "1 batch will be used."
+        log.info(f"`Trainer({name}=1)` was configured so {message}")
+    elif isinstance(batches, float) and batches == 1.0:
+        if name == "limit_train_batches":
+            message = "100% of the batches per epoch will be used."
+        elif name == "val_check_interval":
+            message = "validation will run at the end of the training epoch."
+        else:
+            message = "100% of the batches will be used."
+        log.info(f"`Trainer({name}=1.0)` was configured so {message}.")
```
Comment (ananthsub) [inline pytorch_lightning/trainer/trainer.py:2667]: rank_zero_info ?

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/trainer/trainer.py (hunk 3 above)
Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:2649]: Yes. IMO these should be moved to the loops that use them.

Comment (carmocca) [reply pytorch_lightning/trainer/trainer.py:2649]: With a similar reasoning to #11845

**Review by awaelchli — APPROVED** [review]
